packages/bull-vault/src/AuctionBull.sol

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity =0.7.6;
+
+// interface
+import {IController} from "squeeth-monorepo/interfaces/IController.sol";
+import {IBullStrategy} from "./interface/IBullStrategy.sol";
+import {IERC20} from "openzeppelin/token/ERC20/IERC20.sol";
+// contract
+import {Ownable} from "openzeppelin/access/Ownable.sol";
+import {UniFlash} from "./UniFlash.sol";
+
+import {console} from "forge-std/console.sol";
+
+
+/**
+ * @notice AuctionBull contract
+ * @author opyn team
+ */
+contract AuctionBull is UniFlash, Ownable {
+    /// @dev the highest delta we can have without rebalancing
+    uint256 public deltaUpper;
+    /// @dev the lowest delta we can have without rebalancing
+    uint256 public deltaLower;
+    /// @dev the highest CR we can have before rebalancing
+    uint256 public crUpper;
+    /// @dev the lowest CR we can have in leverage component before rebalancing
+    uint256 public leverageCrLower;
+
+    /// @dev USDC address
+    address private immutable usdc;
+    /// @dev WETH address
+    address private immutable weth;
+    address private immutable bullStrategy;
+    /// @dev auction manager
+    address public auctionManager;
+
+    /// @dev enum to differentiate between Uniswap swap callback function source
+    enum FLASH_SOURCE {
+        GENERAL_SWAP
+    }
+
+    constructor(address _auctionOwner, address _auctionManager, address _bull, address _factory) UniFlash(_factory) Ownable() {
+        bullStrategy = _bull;
+        weth = IController(IBullStrategy(_bull).powerTokenController()).weth();
+        usdc = IController(IBullStrategy(_bull).powerTokenController()).quoteCurrency();
+        auctionManager = _auctionManager;
+
+        transferOwnership(_auctionOwner);
+    }
+
+    /**
+     * @dev changes the leverage component composition by buying or selling eth
+     */
+    function leverageRebalance(bool _isSellingEth, uint256 _amountIn, uint256 _minAmountOut, uint24 _poolFee) external {
+        if (_isSellingEth) {
+            // Withdraw ETH from collateral
+            IBullStrategy(bullStrategy).withdrawWethFromEuler(_amountIn); 
```

Comment:
If we’re doing this for usdc then maybe just calculate:

usdcToTrade = debt - (crab price)*(crab amount)

Then exactIn or exactOut to get that?

Reply:
Decided in Slack that we just take in `usdcAmount`.